Issue #210694 add hoverforeground and hoverbackground to activity bars

AminSajedian opened the pull request:
This pull request introduces customizable hover colors for the side, top, and bottom activity bars.

The changes address issue #210694 (https://github.com/Microsoft/vscode/issues/210694), which requested the ability to set hoverForeground and hoverBackground for a more visible hover state on activity bar icons. This implementation allows users to define specific colors in their settings to improve the discoverability and accessibility of the UI.

I've added the necessary color token properties to the activity bar's color contribution points to enable this feature. @benibenj (https://github.com/benibenj)

Testing:
I have verified that the new color settings work as expected and do not introduce any regressions. The hover state now correctly reflects the colors specified in the user's settings.


Modified Files Summary

1. Theme Color Definition [src/vs/workbench/common/theme.ts]
- Define new theme color tokens:
  - `ACTIVITY_BAR_HOVER_FOREGROUND`
  - `ACTIVITY_BAR_HOVER_BACKGROUND`
  - `ACTIVITY_BAR_TOP_HOVER_FOREGROUND`
  - `ACTIVITY_BAR_TOP_HOVER_BACKGROUND`

2. Activity Bar: Theming + Wiring [src/vs/workbench/browser/parts/activitybar/activitybarPart.ts]
- Register new hover tokens in theme registry
- Map hover tokens to CSS variables for stylesheet consumption
- Wire hover colors into composite bar configuration

3. Activity Bar CSS (Hover Styles) [src/vs/workbench/browser/parts/activitybar/media/activityaction.css]
- Apply hover styles using new CSS variables:
  - `--vscode-activityBar-hoverForeground`
  - `--vscode-activityBar-hoverBackground`
- Integrate hover states with existing active/focus styling

4. Sidebar: Theming Integration [src/vs/workbench/browser/parts/sidebar/sidebarPart.ts]
- Reference top-position hover tokens for consistency:
  - `ACTIVITY_BAR_TOP_HOVER_FOREGROUND`
  - `ACTIVITY_BAR_TOP_HOVER_BACKGROUND`
- Ensure sidebar hover behavior aligns with activity bar when positioned at top or bottom

5. Sidebar CSS (Hover Styles) [src/vs/workbench/browser/parts/sidebar/media/sidebarpart.css]
- Apply hover styling using activity bar CSS variables
- Maintain visual consistency between sidebar and activity bar hover states

6. Stylelint Allowlist [build/lib/stylelint/vscode-known-variables.json]
- Add CSS custom properties to prevent lint errors:
  - `--vscode-activityBar-hoverForeground`
  - `--vscode-activityBar-hoverBackground`

Usage
user's settings:

settings.json
```
{
  "workbench.activityBar.location": "default", // "top" or "bottom"
  "workbench.colorCustomizations": {
    // Left Activity Bar 
    "activityBar.hoverForeground": "#ff0000", 
    "activityBar.hoverBackground": "#00ff00",

    // Top Activity Bar 
    "activityBarTop.hoverForeground": "#0000ff", 
    "activityBarTop.hoverBackground": "#ffff00" 
    } 
}
```

AminSajedian committed:
Add hover colors for side/top/bottom activity bar in theme and styles

Modified Files Summary

1. Theme Color Definition
[src/vs/workbench/common/theme.ts]
- Define new theme color tokens:
  - `ACTIVITY_BAR_HOVER_FOREGROUND`
  - `ACTIVITY_BAR_HOVER_BACKGROUND`
  - `ACTIVITY_BAR_TOP_HOVER_FOREGROUND`
  - `ACTIVITY_BAR_TOP_HOVER_BACKGROUND`

2. Activity Bar: Theming + Wiring
[src/vs/workbench/browser/parts/activitybar/activitybarPart.ts]
- Register new hover tokens in theme registry
- Map hover tokens to CSS variables for stylesheet consumption
- Wire hover colors into composite bar configuration

3. Activity Bar CSS (Hover Styles)
[src/vs/workbench/browser/parts/activitybar/media/activityaction.css]
- Apply hover styles using new CSS variables:
  - `--vscode-activityBar-hoverForeground`
  - `--vscode-activityBar-hoverBackground`
- Integrate hover states with existing active/focus styling

4. Sidebar: Theming Integration
[src/vs/workbench/browser/parts/sidebar/sidebarPart.ts]
- Reference top-position hover tokens for consistency:
  - `ACTIVITY_BAR_TOP_HOVER_FOREGROUND`
  - `ACTIVITY_BAR_TOP_HOVER_BACKGROUND`
- Ensure sidebar hover behavior aligns with activity bar when positioned at top or bottom

5. Sidebar CSS (Hover Styles)
`src/vs/workbench/browser/parts/sidebar/media/sidebarpart.css`
- Apply hover styling using activity bar CSS variables
- Maintain visual consistency between sidebar and activity bar hover states

6. Stylelint Allowlist
[build/lib/stylelint/vscode-known-variables.json]
- Add CSS custom properties to prevent lint errors:
  - `--vscode-activityBar-hoverForeground`
  - `--vscode-activityBar-hoverBackground`

Usage

settings.json
{
  "workbench.activityBar.location": "default", // "top"
  "workbench.colorCustomizations": {
    // Left Activity Bar
    "activityBar.hoverForeground": "#ff0000",
    "activityBar.hoverBackground": "#00ff00",

    // Top Activity Bar
    "activityBarTop.hoverForeground": "#0000ff",
    "activityBarTop.hoverBackground": "#ffff00"
  }
}

AminSajedian committed:
Merge branch 'main' into issue-210694-add-hoverforeground-and-hoverbackground-to-activity-bars

AminSajedian commented:
@microsoft-github-policy-service agree

benibenj committed:
Merge branch 'main' into issue-210694-add-hoverforeground-and-hoverbackground-to-activity-bars

benibenj reviewed (changes requested):
Thanks for creating the PR, this looks mostly good, however, we register default colors in the `theme.ts` file instead of directly in the `.css` files. This allows us to get the color for a theme color through an API

AminSajedian committed:
Merge branch 'microsoft:main' into issue-210694-add-hoverforeground-and-hoverbackground-to-activity-bars

AminSajedian committed:
refactor: move default colors to theme.ts and remove CSS fallbacks in activityaction.css and sidebarpart.css, per feedback.

AminSajedian commented:
Thanks for the helpful feedback! I’ve updated the PR to register default colors in [src/vs/workbench/common/theme.ts] instead of using CSS fallbacks.

Changes:
- Registered defaults in [src/vs/workbench/common/theme.ts]:
  - `activityBar.hoverForeground` → `ACTIVITY_BAR_FOREGROUND`
  - `activityBar.hoverBackground` → `ACTIVITY_BAR_BACKGROUND`
  - `activityBarTop.hoverForeground` → `ACTIVITY_BAR_TOP_FOREGROUND`
  - `activityBarTop.hoverBackground` → `ACTIVITY_BAR_TOP_BACKGROUND`
  
- Updated CSS to reference variables directly without fallbacks:
  - [src/vs/workbench/browser/parts/activitybar/media/activityaction.css]
  - [src/vs/workbench/browser/parts/sidebar/media/sidebarpart.css]

This aligns with the guidance to source defaults from [src/vs/workbench/common/theme.ts] so they are available through the theme API. Please let me know if you’d like any adjustments to the chosen defaults.

AminSajedian committed:
Merge branch 'main' into issue-210694-add-hoverforeground-and-hoverbackground-to-activity-bars

benibenj commented:
Reverted this PR as I lead to this bug

![activity-bar_360](https://github.com/user-attachments/assets/ff8d2ca5-d62b-4bb4-8c09-f9db7129f52a)
